### PR TITLE
TDL-19527 added backoff for 5xx error and added custom error message when only organization is passed in config

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change
- Added backoff for 5xx errors
- Added custom error messages when only organization is passed in config.

# Manual QA steps
 - Verify that the tap backoffs for 5 times for 5xx error.
 - Verify the tap throws proper message when repo name not found in config
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_github/client.py [hunk 1]
```diff
@@ -159,7 +165,7 @@ def get_request_timeout(self):
     # pylint: disable=dangerous-default-value
     # During 'Timeout' error there is also possibility of 'ConnectionError',
     # hence added backoff for 'ConnectionError' too.
-    @backoff.on_exception(backoff.expo, (requests.Timeout, requests.ConnectionError), max_tries=5, factor=2)
+    @backoff.on_exception(backoff.expo, (requests.Timeout, requests.ConnectionError, Server5xxError), max_tries=5, factor=2)
```
Comment (dbshah1212) [inline tap_github/client.py:168]: Error handling for 429 is missing.

File: tap_github/client.py [hunk 2]
```diff
@@ -227,7 +233,14 @@ def extract_repos_from_config(self):
...
 
-        orgs_with_all_repos = list(filter(lambda x: x.split('/')[1] == '*', repo_paths))
+        orgs_with_all_repos = []
+        for each_repo in repo_paths:
+            split_repo_path = each_repo.split('/')
+            if len(split_repo_path) > 1:
+                if split_repo_path[1] == '*':
+                    orgs_with_all_repos.append(each_repo)
+            else:
+                raise Exception("Repository name not found.")
```
Comment (dbshah1212) [inline tap_github/client.py:243]: Use GithubException

File: tap_github/client.py [hunk 3]
```diff
@@ -227,7 +233,14 @@ def extract_repos_from_config(self):
         """
         repo_paths = list(filter(None, self.config['repository'].split(' ')))
 
-        orgs_with_all_repos = list(filter(lambda x: x.split('/')[1] == '*', repo_paths))
+        orgs_with_all_repos = []
+        for each_repo in repo_paths:
```
Comment (dbshah1212) [inline tap_github/client.py:237]: Add code comments

File: tests/unittests/test_exception_handling.py [hunk 4]
```diff
@@ -173,9 +173,19 @@ def test_409_error(self, mocked_parse_args, mocked_request, mock_verify_access):
...
 
+    def test_501_error(self, mocked_parse_args, mocked_request, mock_verify_access):
+        mocked_request.return_value = get_response(501, raise_error = True)
+        test_client = GithubClient(self.config)
+
+        with self.assertRaises(tap_github.client.Server5xxError) as e:
+            test_client.authed_get("", "")
+
+        # Verifying the message formed for the custom exception
+        self.assertEqual(str(e.exception), "HTTP-error-code: 501, Error: Unknown Error")
```
Comment (dbshah1212) [inline tests/unittests/test_exception_handling.py:184]: Add assertion of retry.

File: tests/unittests/test_exception_handling.py [hunk 5]
```diff
@@ -173,9 +173,19 @@ def test_409_error(self, mocked_parse_args, mocked_request, mock_verify_access):
         # Verifying the message formed for the custom exception
         self.assertEqual(str(e.exception), "HTTP-error-code: 409, Error: The request could not be completed due to a conflict with the current state of the server.")
 
+    def test_501_error(self, mocked_parse_args, mocked_request, mock_verify_access):
```
Comment (dbshah1212) [inline tests/unittests/test_exception_handling.py:176]: Add comment

File: tap_github/client.py (hunk 2 above)
Comment (savan-chovatiya) [inline tap_github/client.py:243]: Also, provide the wrong repo_path in an error message.

Comment (namrata270998) [reply tap_github/client.py]: Used GithubException

File: tap_github/client.py (hunk 3 above)
Comment (namrata270998) [reply tap_github/client.py:241]: Added code comments

File: tests/unittests/test_exception_handling.py (hunk 4 above)
Comment (namrata270998) [reply tests/unittests/test_exception_handling.py:187]: Was having an assertion in timeout unittest, however combined in the same and removed that one

File: tests/unittests/test_exception_handling.py (hunk 5 above)
Comment (namrata270998) [reply tests/unittests/test_exception_handling.py:176]: added comments

File: tap_github/client.py [hunk 6]
```diff
@@ -231,7 +241,29 @@ def extract_repos_from_config(self):
         """
         repo_paths = list(filter(None, self.config['repository'].split(' ')))
 
-        orgs_with_all_repos = list(filter(lambda x: x.split('/')[1] == '*', repo_paths))
+        visited = set()
```
Comment (savan-chovatiya) [inline tap_github/client.py:244]:
```suggestion
        unique_repos = set()
```

File: tap_github/client.py [hunk 7]
```diff
@@ -231,7 +241,29 @@ def extract_repos_from_config(self):
...
+        orgs_with_all_repos = []
+        repos_with_errors = []
+        for each_repo in repo_paths:
+            # Split the repo_path by `/`.
+            split_repo_path = each_repo.split('/')
+            if len(split_repo_path) > 1 and split_repo_path[1] != '' and split_repo_path[0] != '':
+                if split_repo_path[1] == '*':
+                    orgs_with_all_repos.append(each_repo)
+            else:
+                # If `/` not found or repo name not found, append the repo_path in the repos_with_errors
```
Comment (savan-chovatiya) [inline tap_github/client.py:261]: Also, add like `Organization not found` too.

File: tap_github/client.py [hunk 8]
```diff
@@ -231,7 +241,29 @@ def extract_repos_from_config(self):
         """
         repo_paths = list(filter(None, self.config['repository'].split(' ')))
 
-        orgs_with_all_repos = list(filter(lambda x: x.split('/')[1] == '*', repo_paths))
+        visited = set()
+        # Insert the duplicate repos found in the config repo_paths into duplicates
+        duplicates = [x for x in repo_paths if x in visited or (visited.add(x) or False)]
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_github/client.py]: Updated the code

File: tap_github/client.py (hunk 6 above)
Comment (namrata270998) [reply tap_github/client.py]: Updated the code

File: tap_github/client.py (hunk 7 above)
Comment (namrata270998) [reply tap_github/client.py]: Updated the comment

File: tap_github/client.py [hunk 9]
```diff
@@ -231,7 +241,29 @@ def extract_repos_from_config(self):
...
+        if duplicates:
+            LOGGER.warn('Duplicate repositories found: {} and will be synced only once.'.format(duplicates))
+
+        repo_paths = list(set(repo_paths))
+
+        orgs_with_all_repos = []
+        repos_with_errors = []
+        for each_repo in repo_paths:
+            # Split the repo_path by `/`.
```
Comment (dbshah1212) [inline tap_github/client.py:255]: Complete the sentence by providing the reason of the split.

File: tap_github/client.py [hunk 10]
```diff
@@ -231,7 +241,29 @@ def extract_repos_from_config(self):
...
+            if len(split_repo_path) > 1 and split_repo_path[1] != '' and split_repo_path[0] != '':
+                if split_repo_path[1] == '*':
+                    orgs_with_all_repos.append(each_repo)
+            else:
+                # If `/` not found or repo name not found, append the repo_path in the repos_with_errors
+                repos_with_errors.append(each_repo)
+        
+        # If any repos found in repos_with_errors, raise an exception
+        if repos_with_errors:
+            raise GithubException("Please provide proper organization/repository: {}".format(repos_with_errors))
```
Comment (dbshah1212) [inline tap_github/client.py:266]:
```suggestion
            raise GithubException("Please provide valid organization/repository for : {}".format(repos_with_errors))
```

File: tests/unittests/test_exception_handling.py [hunk 11]
```diff
@@ -191,9 +191,41 @@ def test_409_error(self, mocked_parse_args, mocked_request, mock_verify_access):
...
+        mocked_request.return_value = get_response(501, raise_error = True)
+        test_client = GithubClient(self.config)
+
+        with self.assertRaises(tap_github.client.Server5xxError) as e:
+            test_client.authed_get("", "")
+
+        # Verifying the message formed for the custom exception
+        self.assertEqual(str(e.exception), "HTTP-error-code: 501, Error: Unknown Error")
+
+        # Verify that we backoff 5 times
```
Comment (dbshah1212) [inline tests/unittests/test_exception_handling.py:207]:
```suggestion
        # Verify 5 times backoff
```

File: tap_github/client.py [hunk 12]
```diff
@@ -231,7 +241,29 @@ def extract_repos_from_config(self):
...
+        # Insert the duplicate repos found in the config repo_paths into duplicates
+        duplicate_repos = [x for x in repo_paths if x in unique_repos or (unique_repos.add(x) or False)]
+        if duplicate_repos:
+            LOGGER.warning("Duplicate repositories found: %s and will be synced only once.", duplicate_repos)
+
+        repo_paths = list(set(repo_paths))
+
+        orgs_with_all_repos = []
+        repos_with_errors = []
+        for each_repo in repo_paths:
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_github/client.py]: Updated the var name

File: tap_github/client.py (hunk 9 above)
Comment (namrata270998) [reply tap_github/client.py]: Updated the comment

File: tap_github/client.py (hunk 10 above)
Comment (namrata270998) [reply tap_github/client.py]: Updated the error message

File: tests/unittests/test_exception_handling.py (hunk 11 above)
Comment (namrata270998) [reply tests/unittests/test_exception_handling.py]: Updated the comment

File: tap_github/client.py [hunk 13]
```diff
@@ -231,7 +241,29 @@ def extract_repos_from_config(self):
         """
         repo_paths = list(filter(None, self.config['repository'].split(' ')))
 
-        orgs_with_all_repos = list(filter(lambda x: x.split('/')[1] == '*', repo_paths))
+        unique_repos  = set()
```
Comment (dbshah1212) [inline tap_github/client.py:244]:
```suggestion
        unique_repos = set()
```

Comment (namrata270998) [reply tap_github/client.py]: Updated the code

File: tap_github/client.py [hunk 14]
```diff
@@ -231,7 +241,29 @@ def extract_repos_from_config(self):
...
+
+        repo_paths = list(set(repo_paths))
+
+        orgs_with_all_repos = []
+        repos_with_errors = []
+        for repo in repo_paths:
+            # Split the repo_path by `/` as we are passing org/repo_name in the config.
+            split_repo_path = repo.split('/')
+            if len(split_repo_path) > 1 and split_repo_path[1] != '' and split_repo_path[0] != '':
+                if split_repo_path[1] == '*':
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_github/client.py:261]: Added comments

**Review by dbshah1212 — APPROVED** [review]

**Pull request merged**